File: 강의 자료/02-알고리즘/09~11-그래프 탐색/문제별 코드/14502-연구소/solution.py
import sys
from collections import deque
si = sys.stdin.readline
n, m = map(int, si().split())
a = [list(map(int, si().split())) for _ in range(n)]
blank = [(i, j) for i in range(n) for j in range(m) if a[i][j] == 0]
ans = 0

def bfs():
    global ans
    cnt = 0
    visit = [[False] * m for _ in range(n)]
    dir = [[1, 0], [0, 1], [-1, 0], [0, -1]]

    Q = deque()
    for i in range(n):
        for j in range(m):
            if a[i][j] == 2:
                visit[i][j] = True
                Q.append(i)
                Q.append(j)

    while Q:
        x = Q.popleft()
        y = Q.popleft()
        for dx, dy in dir:
            nx, ny = x + dx, y + dy
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue
            if a[nx][ny] != 0:
                continue
            if visit[nx][ny]:
                continue
            visit[nx][ny] = True
            Q.append(nx)
            Q.append(ny)

    for i in range(n):
        for j in range(m):
            if a[i][j] == 0 and not visit[i][j]:
                cnt += 1
    
    ans = max(ans, cnt)


def dfs(idx, selected_cnt):
    if selected_cnt == 3:
        bfs()
        return
    
    if idx == len(blank):
        return
    
    a[blank[idx][0]][blank[idx][1]] = 1
    dfs(idx + 1, selected_cnt + 1)

    a[blank[idx][0]][blank[idx][1]] = 0
    dfs(idx + 1, selected_cnt)

dfs(0, 0)
print(ans)

File: 강의 자료/02-알고리즘/09~11-그래프 탐색/문제별 코드/14502-연구소/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

#define NM 11

int N, M, B, A[NM][NM], blank[NM * NM][2], ans;
bool visit[NM][NM];
int dir[4][2] = { {1,0},{0,1},{-1,0},{0,-1} };

void input() {
	cin >> N >> M;
	for (int i = 1; i <= N; i++)
		for (int j = 1; j <= M; j++)
			cin >> A[i][j];
}

#include <queue>
void bfs() {
	queue<int> Q;

	// 모든 바이러스가 시작점으로 가능하니까, 전부 큐에 넣어준다.
	for (int i = 1; i <= N; i++) {
		for (int j = 1; j <= M; j++) {
			visit[i][j] = false;
			if (A[i][j] == 2) {
				Q.push(i);
				Q.push(j);
				visit[i][j] = true;
			}
		}
	}

	// BFS 과정
	while (!Q.empty()) {
		int x = Q.front(); Q.pop();
		int y = Q.front(); Q.pop();
		for (int k = 0; k < 4; k++) {
			int nx = x + dir[k][0], ny = y + dir[k][1];
			if (nx < 1 || ny < 1 || nx > N || ny > M) continue;
			if (A[nx][ny] != 0) continue;
			if (visit[nx][ny]) continue;
			visit[nx][ny] = true;
			Q.push(nx);
			Q.push(ny);
		}
	}

	// 탐색이 종료된 시점이니, 안전 영역의 넓이를 계산하고, 정답을 갱신한다.
	int cnt = 0;
	for (int i = 1; i <= N; i++) for (int j = 1; j <= M; j++) if (A[i][j] == 0 && !visit[i][j]) cnt++;
	ans = max(ans, cnt);
}

void dfs(int idx, int selected_cnt) {
	if (selected_cnt == 3) {
		bfs();
		return;
	}

	if (idx > B) return;
	
	A[blank[idx][0]][blank[idx][1]] = 1;
	dfs(idx + 1, selected_cnt + 1);

	A[blank[idx][0]][blank[idx][1]] = 0;
	dfs(idx + 1, selected_cnt);
}

void pro() {
	for (int i=1;i<=N;i++)
		for (int j=1;j<=M;j++)
			if (A[i][j] == 0) {
				B++;
				blank[B][0] = i;
				blank[B][1] = j;
			}
	dfs(1, 0);
	cout << ans;
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}